lib/modules/versioning/hermit/index.spec.ts

```diff
@@ -0,0 +1,276 @@
+      ${'17.2.3a1'}     | ${[17, 2, 3]}     | ${undefined}  | ${'a1'}
+      ${'17.2.3-foo'}   | ${[17, 2, 3]}     | ${undefined}  | ${'foo'}
+      ${'17.2.3+m1'}    | ${[17, 2, 3]}     | ${'m1'}       | ${undefined}
+      ${'17.2.3_40+m1'} | ${[17, 2, 3, 40]} | ${'m1'}       | ${undefined}
+      ${'@17'}          | ${[17]}           | ${undefined}  | ${undefined}
+      ${'@17.2'}        | ${[17, 2]}        | ${undefined}  | ${undefined}
+      ${'@17.2.3_40'}   | ${[17, 2, 3, 40]} | ${undefined}  | ${undefined}
+    `(
+      'getReleases for "$version"',
+      ({ version, release, compatibility, prerelease }) => {
+        const res = versioning._parseVersion(version);
+        if (res === null) {
+          throw new Error('res is null');
+        }
+        expect(res.release).toStrictEqual(release);
+        expect(res.compatibility).toStrictEqual(compatibility);
+        expect(res.prerelease).toStrictEqual(prerelease);
+      }
+    );
+
+    it('non semver channel will receive a null', () => {
+      expect(versioning._parseVersion('@stable')).toBeNull();
+    });
+  });
+
+  describe('_isChannel', () => {
+    test.each`
+      version                     | expected
+      ${'1'}                      | ${false}
+      ${'1rc1'}                   | ${false}
+      ${'1-foo'}                  | ${false}
+      ${'1+bar'}                  | ${false}
+      ${'1.2'}                    | ${false}
+      ${'1.2-foo'}                | ${false}
+      ${'1.2+bar'}                | ${false}
+      ${'1.2.3'}                  | ${false}
+      ${'1.2.3rc1'}               | ${false}
+      ${'1.2.3-foo'}              | ${false}
+      ${'1.2.3+bar'}              | ${false}
+      ${'17.0.1_12'}              | ${false}
+      ${'17.0.1_12+m1'}           | ${false}
+      ${'17.0.1_12+m1'}           | ${false}
+      ${'11.0.11_9-zulu11.48.21'} | ${false}
+      ${'1.2-kotlin.3'}           | ${false}
+      ${'@1'}                     | ${true}
+      ${'@1.2'}                   | ${true}
+      ${'@1.2.3'}                 | ${true}
+      ${'@latest'}                | ${true}
+      ${'@stable'}                | ${true}
+    `('isChannel("$version") === $expected', ({ version, expected }) => {
+      expect(!!versioning._isChannel(version)).toBe(expected);
+    });
+  });
+
+  describe('_isSemverChannel', () => {
+    test.each`
+      version      | expected
+      ${'@1'}      | ${true}
+      ${'@1.2'}    | ${true}
+      ${'@1.2.3'}  | ${true}
+      ${'@latest'} | ${false}
+      ${'@stable'} | ${false}
+    `('isSemverChannel("$version") === $expected', ({ version, expected }) => {
+      expect(!!versioning._isSemverChannel(version)).toBe(expected);
+    });
+  });
+
+  describe('isStable', () => {
+    test.each`
+      version      | expected
+      ${'1'}       | ${true}
+      ${'1.2'}     | ${true}
+      ${'@1'}      | ${false}
+      ${'@1.2'}    | ${false}
+      ${'@1.2.3'}  | ${false}
+      ${'@latest'} | ${false}
```

Comment:
Is `@latest` considered stable? I'm not 100% sure from the documentation[^1]:

> Additionally, Hermit will create several synthetic channels which are checked for updates every 24h:
>
> 1. A `@latest` channel pointing at the most recent non-pre-release version.
> 2. A `@<MAJOR>` channel for every major version.
> 3. A `@<MAJOR>.<MINOR>` channel for every major+minor version.
>
> This allows projects to pin to stable releases.

Does "stable releases" refer to all three, or only the last two?

[^1]: https://cashapp.github.io/hermit/packaging/reference/#versions

Reply:
So, the "stable releases" in the section you referred to includes all 3 of the bullet points. It is saying all those 3 channels will be only pointing to "stable releases", comparing to non-stable pre-releases ( e.g. v1.2.3-beta2) .

Hermit packages using channels will automatically update before run whenever a new version is available through `hermit-packages`. 

An example of this is: in CI environment, if a CI tool is installed with the `@latest` channel, it might update in an arbitrary time which breaks the pipeline that had passed before.

Hence when working with Renovate, channels are not considered `stable`.

Reply:
🤔 It looks like a channel should be handled as a range. Whould the datasource ever return a channel? or are those only read by the manager?

Reply:
☝️?

Reply:
Datasource would only return the name of the channel. Not the details. 

We are going to slowly discourage the use of Channel due to its undetermined characteristics and prefer the use of dependency managers like renovate to manage the package versions.